### PR TITLE
Upload template to s3

### PR DESCRIPTION
**Pull request by laurence-bird**

### CODE REVIEW

File: cloudformation.yaml [hunk 1]
```diff
@@ -77,11 +77,11 @@ Resources:
     Type: AWS::ElasticLoadBalancingV2::LoadBalancer
     Properties:
       LoadBalancerAttributes:
-        - Key: access_logs.s3.enabled
+        - Key: access_logs.aws.s3.enabled
```
Comment (cb372) [inline cloudformation.yaml]: These attribute names were correct before, so please don't change them. See e.g. [here](http://docs.aws.amazon.com/AWSJavaSDK/latest/javadoc/com/amazonaws/services/elasticloadbalancingv2/model/LoadBalancerAttribute.html#setKey-java.lang.String-)

File: conf/prd.conf [hunk 2]
```diff
@@ -0,0 +1,31 @@
+play.crypto.secret = "ADHFSDJGHSDFKJSFDDS23456rterwdfghterfds"
```
Comment (cb372) [inline conf/prd.conf]: This needs to be a secret (the clue is in the name), so we'll need to inject it using credstash. Should do the same for UAT. The one in `application.conf` doesn't matter as it's only used when running locally.

File: project/plugins.sbt [hunk 3]
```diff
@@ -1,4 +1,5 @@
 addSbtPlugin("com.typesafe.play" % "sbt-plugin" % "2.5.10")
-
+addSbtPlugin("com.tapad" % "sbt-docker-compose" % "1.0.13")
 resolvers += Resolver.bintrayIvyRepo("ovotech", "sbt-plugins")
-addSbtPlugin("com.ovoenergy" % "sbt-comms-packaging" % "0.0.2")
+addSbtPlugin("com.ovoenergy" % "sbt-comms-packaging" % "0.0.3-SNAPSHOT")
```
Comment (cb372) [inline project/plugins.sbt]: Did you release a new version of sbt-comms-packaging? If it's working, please release it and use the release version.

File: test/util/LocalDynamoDB.scala [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+  }
+
+  def usingTable[T](client: AmazonDynamoDB)(tableName: String)(attributeDefinitions: (Symbol, ScalarAttributeType)*)(
+    thunk: => T
+  ): Unit = {
+    withTable(client)(tableName)(attributeDefinitions: _*)(thunk)
+    ()
+  }
+
+  def withTableWithSecondaryIndex[T](client: AmazonDynamoDB)(tableName: String, secondaryIndexName: String)
```
Comment (cb372) [inline test/util/LocalDynamoDB.scala]: Could this function reuse `createTableWithSecondaryIndex`?

File: cloudformation.yaml (hunk 1 above)
Comment (laurence-bird) [inline cloudformation.yaml]: oops, think the IDE did this during a refactor, well spotted!

File: app/AppLoader.scala [hunk 5]
```diff
@@ -1,14 +1,14 @@
 import play.api.libs.logback.LogbackLoggerConfigurator
-import play.api.{Application, ApplicationLoader}
+import play.api.{Application, ApplicationLoader, Logger}
 import play.api.ApplicationLoader.Context
 
 
-class AppLoader extends ApplicationLoader {
+class
+AppLoader extends ApplicationLoader {
```
Comment (cb372) [inline app/AppLoader.scala]: Weird newline here. Actually you can just revert this whole file

File: app/aws/dynamo/Dynamo.scala [hunk 6]
```diff
@@ -30,33 +43,41 @@ object Dynamo {
...
 
   def writeEvent[A: WritableEvent](event: A) = ReaderT[Try, AmazonDynamoDB, UpdateItemResult] { dynamo =>
-    val ev = WritableEvent[A]
+    val ev: WritableEvent[A] = WritableEvent[A]
+
 
     // The triggered event contains the "correct" timestamp for the comm, so it should be set unconditionally.
     // For any other type of event, set a provisional timestamp if one doesn't exist yet
     // (because the Triggered event hasn't arrived yet)
+    Logger.info(s"Writing event to table ${table.name}")
```
Comment (cb372) [inline app/aws/dynamo/Dynamo.scala]: Not sure how useful this log message is. We don't care about the table name, and writing an INFO log for every single write to Dynamo could be quite noisy.

Comment (cb372) [inline app/aws/dynamo/Dynamo.scala]: If you want to keep noisy logging temporarily, I would include the event type and trace token in the log message

File: app/models/DynamoFormats.scala [hunk 7]
```diff
@@ -35,6 +36,39 @@ object DynamoFormats {
     case EmailStatus.Opened => "Opened"
   }
 
+  implicit val errorCodeFormat = DynamoFormat.coercedXmap[ErrorCode, String, MatchError]{
```
Comment (cb372) [inline app/models/DynamoFormats.scala]: If we upgrade to Scanamo 0.9, we can delete all these boilerplate-y DynamoFormats for enums 🎉 

https://github.com/guardian/scanamo/pull/84

File: app/repo/EmailBodyRepo.scala [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+import aws.s3.AmazonS3ClientWrapper._
+
+object EmailBodyRepo {
+
+  type BucketName   = String
+  type ResourceKey  = String
+
+  type ErrorOr[A] = Either[String, A]
+
+  case class UploadError(msg: String, contents: ArchivedContentsType)
```
Comment (cb372) [inline app/repo/EmailBodyRepo.scala]: This class doesn't seem to be used, and `ArchivedContentsType` is also not used anywhere except here

File: app/repo/EmailBodyRepo.scala [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+    val htmlBodyUrl: ErrorOr[ResourceUrl] = fileUploader(htmlFileDetails)
+
+    val textBodyUrl = event.textBody.map{ txtBody =>
+      val fileDetails = S3FileDetails(txtBody, s"$emailTextFolder/${event.internalMetadata.internalTraceToken}", s3Config.bucketName)
+      fileUploader(fileDetails)
+    }
+
+    (htmlBodyUrl, textBodyUrl) match {
+      case (html, Some(text)) =>  event.copy(htmlBody = html.merge, textBody = Some(text.merge))
+      case (html, _)          =>  event.copy(htmlBody = html.merge)
```
Comment (cb372) [inline app/repo/EmailBodyRepo.scala]: Nice use of `merge`! I always forget about that method

File: test/ServiceTestIT.scala [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+  val composedEmailProducer = KafkaProducer(KafkaProducerConf(new StringSerializer, avroSerializer[ComposedEmail], kafkaHosts))
+  val dynamoClient          = LocalDynamoDB.client(dynamoUrl)
+
+  createTable()
+
+  it should "Consume composedEmail events and write appropriate data to dynamodb" taggedAs DockerComposeTag in {
+    initialiseS3Bucket()
+    createTopics()
+    // Wait for aws.dynamo to create table
+    Thread.sleep(2000)
```
Comment (cb372) [inline test/ServiceTestIT.scala]: Do you still need this sleep now that you've added the polling for table creation?

File: app/models/DynamoFormats.scala (hunk 7 above)
Comment (laurence-bird) [inline app/models/DynamoFormats.scala]: 🔥

**Pull request merged**